wine*: Clean up Portfiles, dependencies and add some patches

casr opened the pull request:
###### Description

After trying to help out with [Trac#53244](https://trac.macports.org/ticket/53244) I noticed that the Portfile had fallen behind with respect to how later versions of Wine were being built. As I got through it, I noticed some unnecessary dependencies and factored out X11 as a variant (there is now a native “Mac Driver”). A couple of patches have been added to fix various behaviour issues on macOS.

There are many small commits below as I thought with this many changes as much documenting would be wise so others can understand the reasoning behind each small change.

Some unresolved things:

  * As X11 is no longer a requirement, I believe that Wine would be better suited in the category ‘emulators’ (Yes, I’m aware what Wine stands for... ;)
  * Check on macOS versions below 10.12. Can buildbot take care of this?
  * X11 still remains on by default but maybe this is unnecessary.
  * When trace mode is enabled, MacPorts picks up that the ports will try and access `git` and that it is not listed as a build dependency. It seems that it is just used by Wine for outputting the version information but it is able to fallback to a predefined variable without Git, therefore, I am just ignoring this.

/cc relevant maintainers @ryandesign @jyrkiwahlstedt @jeremyhu 

###### Tested on
macOS 10.12.3
Xcode 8.2.1

###### Verification
Have you
- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
- [x] checked your Portfile with `port lint`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

casr commented:
Resolved the upstream conflict with `wine-devel`'s version bump.

casr commented:
Resolved the upstream conflict with `wine-devel` and `wine-crossover's version bump.

@ryandesign @jyrkiwahlstedt if you have any thoughts about getting this merged I would be happy to fix things up.

mojca commented on x11/wine-crossover/Portfile:
```diff
@@ -99,32 +99,18 @@ configure.install           ${worksrcpath}/tools/install-sh
-                            --with-cups \
-                            --with-curses \
-                            --with-fontconfig \
-                            --with-freetype \
```
I probably understand your concerns, but why is this particular change needed?

MacPorts generally tries to create as reproducible builds as possible. Though that usually means using lots of `--without-something` and `--disable-somethingelse` to make sure that libraries are not opportunistically linked. I guess the flags above don't make any real difference, but ...

mojca commented on x11/wine-crossover/Portfile:
```diff
@@ -118,26 +108,49 @@ configure.args              --without-alsa \
+                              --x-lib=${prefix}/lib
+}
+
+default_variants            +x11
```
I would be in favour of not making this a default variant (if it works flawlessly). But see also problems reported in https://trac.macports.org/ticket/49800.

casr replied:
I thought it could still be useful to have it be built by the buildbot. The MacDriver is still the preferred way that it launches so you would have to explicitly ask for the X11 one to load. Of course, if you're building from scratch it is nice to not have to pull in all the X11 dependencies.

Not sure if `+quartz` would be a necessity, at least, not now. Maybe if in the future macOS gets a third output option then perhaps?

mojca replied:
Ah, OK, if `+x11` builds both GUIs, then this is even better. I thought that building with X11 excluded the native GUI and vice versa. In that case keeping X11 makes sense.

mojca commented:
The port wants `perl5.24 +universal` which is most likely not necessary.

casr commented:
> I hope you will also take a look at 64-bit support https://trac.macports.org/ticket/53651 ;)

In case you missed it, I made the patch for that to work ;) I was waiting until this got merged in before cleaning producing a final patch though.

> The port wants perl5.24 +universal which is most likely not necessary.
>
> Full list of dependencies being asked for (I had wine-devel already installed from earlier):
>
>     fftw-3
>     fftw-3-single
>     flac
>     libsndfile
>     openal-soft
>     p5.24-xml-parser
>     perl5.24
>     portaudio
>     pulseaudio
>     soxr
>     speexDSP

Do you mean these are additional direct dependencies that were not in the Porfile? I thought I check that all out but it has been quite a while now so I am not sure!

Finally, thank you for finding and linking all the relevant Trac tickets! I've gone and updated the commit messages :)

mojca commented:
> Do you mean these are additional direct dependencies that were not in the Porfile?

No, what I mean is that I already have `wine-devel` installed, so most dependencies should be installed already. You added a dependency on, say, `openal-soft`, so it's expected that your port pulls in a few additional dependencies (`openal-soft` and some of its dependencies). However I see no reason why `perl5.24` or `p5.24-xml-parser` would have to be built with `+universal` unless wine or one of the libraries actually links against `libperl.dylib`. We need to figure out how to best prevent that. I simply listed all dependencies that were still missing on my machine just for double-checking if all of them make sense.

casr commented:
> However I see no reason why `perl5.24` or `p5.24-xml-parser` would have to be built with `+universal` unless wine or one of the libraries actually links against `libperl.dylib`.

Oh, I see. Yes, perhaps that extra configurability would be nice but I think it may also complicate things a whole lot. Could be tricky!

> And sorry for not pay attention about the 64-bitness (I just went through the tickets).

Haha, not at all! I'm very grateful you found all those tickets I missed. And cross-linking it here should help some others out too :)

mojca commented:
I would also suggest creating subports rather than having to keep three independent ports fully in sync (I happily volunteer to do that), but it's probably best if this PR gets merged first.

ryandesign commented:
> I would also suggest creating subports rather than having to keep three independent ports fully in sync (I happily volunteer to do that), but it's probably best if this PR gets merged first.

I'm not a big fan of subports being used in that manner. There are often differences in development versions, and especially in the crossover fork, that are handled well enough as separate ports.

ryandesign commented:
> However I see no reason why perl5.24 or p5.24-xml-parser would have to be built with +universal unless wine or one of the libraries actually links against libperl.dylib. We need to figure out how to best prevent that.

The only thing that occurs to me is `depends_skip_archcheck` but I don't know if it will actually prevent universal installation of those dependencies. If not, then MacPorts base doesn't have this feature yet.

mojca commented:
`depends_skip_archcheck` is exactly what I wanted to suggest. It didn't work out-of-the-box for me for some reason, but I assume there is some straightforward way. Most likely one of the dependencies should declare that.

mojca commented:
A build on 10.7 breaks some time pretty late in the build process:
```
/opt/local/bin/clang-mp-3.4 -m32 -c -o macro.lex.yy.o macro.lex.yy.c -I. -I../../include -D__WINESRC__ -D_REENTRANT -fPIC -Wall -pipe \
  -fno-strict-aliasing -Wdeclaration-after-statement -Wempty-body -Wignored-qualifiers \
  -Wstrict-prototypes -Wtype-limits -Wvla -Wwrite-strings -Wpointer-arith -fno-omit-frame-pointer \
  -I/opt/local/include -pipe -Os -arch i386 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=0
../../tools/winegcc/winegcc -o winhlp32.exe.so -B../../tools/winebuild -m32 -fasynchronous-unwind-tables \
  -mwindows callback.o hlpfile.o macro.o string.o winhelp.o macro.lex.yy.o winhlp32.res -lshell32 \
  -lcomctl32 -lcomdlg32 -luser32 -lgdi32 ../../libs/port/libwine_port.a -Wb,-dshell32 -Wb,-dcomctl32 \
  -Wb,-dcomdlg32 -L/opt/local/lib -Wl,-headerpad_max_install_names -arch i386
Undefined symbols for architecture i386:
  "_yywrap", referenced from:
      _yylex in macro.lex.yy.o
ld: symbol(s) not found for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)
winegcc: /opt/local/bin/clang-mp-3.4 failed
make[1]: *** [winhlp32.exe.so] Error 2
make[1]: Leaving directory `/path/to/wine-devel/work/wine-2.5/programs/winhlp32'
make: *** [programs/winhlp32] Error 2
make: *** Waiting for unfinished jobs....
```

mojca commented:
Indeed, manually adding `-lfl` ended up in successfully installing wine. Apparently some `Makefile` needs patching somewhere.

I still get random problems, but I don't think it is any worse than it was (there are some tickets on Trac):

```
$ wine test.exe 
Hello world.
$ wine: Unhandled page fault on read access to 0x00000000 at address 0x99bf4101 (thread 0032), starting debugger...
wine client error:32: write: Bad file descriptor
err:ntdll:RtlLeaveCriticalSection section 0x110940 is not acquired
```

Meanwhile wine 2.6 has been released.

Ionic commented:
Just a general piece of advise: be careful with removing dependencies. WINE likes to load libraries dynamically, so unless you know the code and what might be used exactly, it's quite difficult to determine the real dependencies. `otool` won't really help.

casr commented:
@ionic oh no! Has something broken for you?

I read through the autotools code and the current build instructions as well as trying some non-trivial programmes to makes sure everything was working. I assumed that if I had removed too much the issue would be mentioned at the PR stage or very soon after merge and that it would be better to be more strict than loose.

casr commented:
> A build on 10.7 breaks some time pretty late in the build process:
> …
> But manually adding -lfl seems to help.

@mojca Was that broken before this PR? I haven't changed anything directly with that flag…

mojca commented:
No, it wasn't broken before this PR. But that's either a new bug in wget's `Makefile`s or an old bug that has been exposed due to slightly changed `configure` options. It should be reported upstream.

Ionic commented:
No, that was just a general comment. I haven't tried building and using your branch (also because I don't have a complex wine testing suite or the like...)

libflex breakages are probably caused by the updated flex port, since the current release of flex has a bug, but IIRC an important patch was backported so that software using flex shouldn't break any longer.

casr commented:
Resolved the conflict with `x11/wine/Portfile` being bumped to `2.0.1`. I removed the [DXTC-to-ARGB patch](http://bugs.winehq.org/show_bug.cgi?id=14939) as that has been included in `2.0.1`'s list of fixes.

FYI, I haven't done a new build for this.

mojca commented:
@casr: In the meantime @ryandesign updated wine-devel to version 2.6, so this PR needs to be fixed.

@ryandesign: Any thoughts about these two wine-related PRs? I see that flex was reverted to an older version, so maybe my problem is gone as well, I need to check.

casr commented:
@mojca resolved the conflict :)

casr commented:
The only conflicts that are occurring are because of the fact that I bumped the revision number. As this PR seems in it for the long haul, I've just gone ahead and deleted that commit.

It will, of course, need to be reintegrated if this ever gets merged.

ryandesign commented:
I am not particularly involved with GitHub pull requests, but taking a look at this now:

I don't know why you are reformatting the portfile comments.

I don't know why you are removing the comment that most libraries are dlopen'd. Can you provide a citation that shows this is no longer true?

Not using fontconfig seems good, assuming that works. I don't know if anyone will complain or will want a fontconfig variant to be available; if so we can add that later.

Removing `build.target` seems fine.

I guess "removing unnecessary flags" is fine.

The whitespace in the x11 variant does not conform to the modeline.

Updating the description is fine.

Removing the Tiger patch is fine since support for Tiger seems long gone.

Fullscreen flickering patch is fine I guess.

Removing configure flags that disable pulseaudio and udev: If these options/features are no longer supported then that's fine.

I disagree with removing the configure flags that make explicit what options we are enabling or disabling, even if they are the default behavior of the configure script. We want to be explicit, and we need the list of dependencies to match. If upstream decided to turn a feature on or off by default in a new version, we would either have unnecessary dependencies or opportunistically-used dependencies.

Adding openal-soft support is fine if that works.

The notes are hard wrapped; please allow MacPorts to line wrap to the width of the terminal. Also observe the statements about indentation made by the modeline.

The patchfiles should not be renamed. The old "patch-XXX.diff" patchfile naming scheme advocated by `port lint` seems to be falling out of favor, with "XXX.patch" taking its place. We need to update the lint warning to allow either form.

Since this PR changes installed files, a revision increase will be needed.

I'm not merging pull requests but the ports are openmaintainer so someone else can merge it when it's ready.

casr commented:
@ryandesign Thank you for reviewing the commits.

> I don't know why you are reformatting the portfile comments.
>
> I don't know why you are removing the comment that most libraries are dlopen'd. Can you provide a citation that shows this is no longer true?

Agree on both points; I should not have made those changes. I have undone that.


> The whitespace in the x11 variant does not conform to the modeline.

I have pushed it to 4 spaces.

> I disagree with removing the configure flags that make explicit what options we are enabling or disabling, even if they are the default behavior of the configure script. We want to be explicit, and we need the list of dependencies to match. If upstream decided to turn a feature on or off by default in a new version, we would either have unnecessary dependencies or opportunistically-used dependencies.

Okay, so I've reverted that commit but then also added in all the other missing options. I have taken a best guess at the defaults but if there is a better way of establishing what `./configure` eventually decides upon I would love to know it. (Demystifying automake scripts is not fun…). Wine's configure script seems to ignore `--with`/`--without` options on some platforms as well as I think is the case with OpenGL on macOS, for instance.

> The notes are hard wrapped; please allow MacPorts to line wrap to the width of the terminal. Also observe the statements about indentation made by the modeline.

Oops! Fixed.

> The patchfiles should not be renamed. The old "patch-XXX.diff" patchfile naming scheme advocated by `port lint` seems to be falling out of favor, with "XXX.patch" taking its place. We need to update the lint warning to allow either form.

Undone.

casr commented:
Resolved `wine-devel`'s bump to 2.7

casr commented:
> I'm not merging pull requests but the ports are openmaintainer so someone else can merge it when it's ready.

@mojca considering the above, perhaps you would be able to merge this in?

mojca commented:
Can you grant permission for committers to change your branch?

casr commented:
@mojca Thank you for merging this :)
